jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/BufferedContentSink.java

```diff
@@ -0,0 +1,266 @@
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("complete; writing aggregated buffer: {} bytes", aggregatedBuffer.remaining());
+                _flusher.offer(true, aggregatedBuffer.getByteBuffer(), Callback.from(callback, aggregatedBuffer::release));
+            }
+            else
+            {
+                _flusher.offer(true, BufferUtil.EMPTY_BUFFER, callback);
+            }
+        }
+        else if (full)
+        {
+            RetainableByteBuffer aggregatedBuffer = _aggregator.takeRetainableByteBuffer();
+            if (LOG.isDebugEnabled())
+                LOG.debug("writing aggregated buffer: {} bytes", aggregatedBuffer.remaining());
+            _flusher.offer(false, aggregatedBuffer.getByteBuffer(), new Callback.Nested(Callback.from(aggregatedBuffer::release))
+            {
+                @Override
+                public void succeeded()
+                {
+                    super.succeeded();
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("written aggregated buffer, writing remaining of current: {} bytes{}", currentBuffer.remaining(), (last ? " (last write)" : ""));
+                    if (last)
+                        _flusher.offer(true, currentBuffer, callback);
+                    else
+                        aggregateAndFlush(false, currentBuffer, callback);
+                }
+
+                @Override
+                public void failed(Throwable x)
+                {
+                    super.failed(x);
+                    callback.failed(x);
+                }
+            });
+        }
+        else
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("buffer fully aggregated, delaying writing - aggregator: {}", _aggregator);
+            _flusher.offer(callback);
+        }
+    }
+
+    private static class Flusher extends IteratingCallback
+    {
+        private static final ByteBuffer COMPLETE_CALLBACK = BufferUtil.allocate(0);
+
+        private final Content.Sink _sink;
+        private boolean _last;
+        private ByteBuffer _buffer;
+        private Callback _callback;
+        private boolean _lastWritten;
+
+        Flusher(Content.Sink sink)
+        {
+            _sink = sink;
+        }
+
+        void offer(Callback callback)
+        {
+            offer(false, COMPLETE_CALLBACK, callback);
+        }
+
+        void offer(boolean last, ByteBuffer byteBuffer, Callback callback)
+        {
+            if (_callback != null)
+                throw new WritePendingException();
+            _last = last;
+            _buffer = byteBuffer;
+            _callback = callback;
+            iterate();
+        }
+
+        @Override
+        protected Action process()
+        {
+            if (_lastWritten)
+                return Action.SUCCEEDED;
+            if (_callback == null)
+                return Action.IDLE;
+            _lastWritten = _last;
```

Comment:
Shouldn't think only be done if you do the write below?

Reply:
`_last` can only be true if `_buffer != COMPLETE_CALLBACK`. But I moved that assignment within the `if()` block to improve readability.